Source/DafnyStandardLibraries/src/Std/Parsers/Core/Parsers.dfy

```diff
@@ -0,0 +1,819 @@
+  }
+
+  lemma AboutMaybe(input: Input)
+    requires A.Length(input) > 0
+  {
+    reveal FailWith();
+    var u := FailWith<char>("Error")(input);
+    reveal Maybe();
+    assert Maybe<char>(FailWith("Error"))(input)
+        == ParseSuccess(Wrappers.None, input);
+    var committedFailure := ParseFailure(Recoverable, FailureData("", A.Drop(input, 1), Option.None));
+    reveal ResultWith();
+    assert Maybe<char>(ResultWith(committedFailure))(input)
+        == committedFailure.PropagateFailure();
+  }
+
+  /** `Maybe(a)` evaluates `a` on the input, and then
+  - If `a` succeeds, then wraps its result with `Some(...)`
+  - If `a` fails, and the failure is not fatal and did not consume input, succeeds with `None`.
+    If the error is fatal or did consume input, fails with the same failure.*/
+  opaque function Maybe<R>(underlying: Parser<R>): Parser<Option<R>>
+  {
+    (input: Input) =>
+      var u := underlying(input);
+      if u.IsFatalFailure() || (u.IsFailure() && !u.NeedsAlternative(input)) then u.PropagateFailure()
+      else
+      if u.ParseSuccess? then u.Map(result => Option.Some(result))
+      else ParseSuccess(Option.None, input)
+  }
+
+  opaque function ConcatMap<L, R, T>(
+    left: Parser<L>,
+    right: Parser<R>,
+    mapper: (L, R) -> T
+  ) : (p: Parser<T>)
+    // Apply two consecutive parsers consecutively
+    // If both succeed, apply the mapper to the result and return it
+  {
+    (input: Input)
+    =>
+      var (l, remaining) :- left(input);
+      var (r, remaining2) :- right(remaining);
+      ParseSuccess(mapper(l, r), remaining2)
+  }
+
+  opaque function Concat<L, R>(
+    left: Parser<L>,
+    right: Parser<R>
+  ) : (p: Parser<(L, R)>)
+    // Apply two consecutive parsers consecutively
+    // If both succeed, return the pair of the two results
+  {
+    (input: Input) =>
+      var (l, remaining) :- left(input);
+      var (r, remaining2) :- right(remaining);
+      ParseSuccess((l, r), remaining2)
+  }
+
+  opaque function ConcatKeepRight<L, R>(
+    left: Parser<L>,
+    right: Parser<R>
+  ) : (p: Parser<R>)
+    // Return only the result of the right parser if the two parsers match
+  {
+    ConcatMap(left, right, (l, r) => r)
+  }
+
+  opaque function ConcatKeepLeft<L, R>(
+    left: Parser<L>,
+    right: Parser<R>
+  ) : (p: Parser<L>)
+    // Return only the result of the right parser if the two parsers match
+  {
+    ConcatMap(left, right, (l, r) => l)
+  }
+
+  /** Use a function that can print to debug, and possibly modify the parser*/
+  opaque function Debug<R, D>(underlying: Parser<R>, name: string, onEnter: (string, Input) -> D, onExit: (string, D, ParseResult<R>) -> ()): (p: Parser<R>) {
+    (input: Input) =>
+      var debugData := onEnter(name, input);
+      var output := underlying(input);
+      var _ := onExit(name, debugData, output);
+      output
+  }
+
+  // A datatype that avoid a quadratic complexity in concatenating long sequences.
```

Comment:
Why is this necessary given all the backends already do this for sequence concatenation?

Reply:
Doing this resulted in a +430% speedup on a benchmark that was using single-character repetition when using the JavaScript version of the parsers.
I think that the problem is when we do this.

    "a" + "b" + "c" + "d" + ...

it does perform the concatenation in a lazy way, but when the time comes when we want to realize the final strings, two things happen
- The recursion to get the string blows the stack in some cases
- When it does not blow the stack, intermediate strings seem to be concretized even though they are not necessary.